LinkageCheckerMain should not fail upon missing Artifact

```
> Task :checkJavaLinkage
SLF4J: Failed to load class "org.slf4j.impl.StaticLoggerBinder".
SLF4J: Defaulting to no-operation (NOP) logger implementation
SLF4J: See http://www.slf4j.org/codes.html#StaticLoggerBinder for further details.
Exception in thread "main" org.eclipse.aether.RepositoryException: Unresolved artifacts: xerces:xerces-impl:jar:2.6.2 was not resolved. Dependency path: org.apache.beam:beam-sdks-java-extensions-sql-zetasql:jar:2.20.0-SNAPSHOT (compile) > org.apache.beam:beam-sdks-java-extensions-sql:jar:2.20.0-SNAPSHOT (compile) > com.alibaba:fastjson:jar:1.2.49 (compile) > org.apache.cxf:cxf-rt-transports-http:jar:3.1.2 (provided?) > org.apache.cxf:cxf-core:jar:3.1.2 (compile) > cglib:cglib-nodep:jar:2.2.2 (provided) > ant:ant:jar:1.6.2 (compile?) > xerces:xerces-impl:jar:2.6.2 (compile?), xml-apis:xml-apis:jar:2.6.2 was not resolved. Dependency path: org.apache.beam:beam-sdks-java-extensions-sql-zetasql:jar:2.20.0-SNAPSHOT (compile) > org.apache.beam:beam-sdks-java-extensions-sql:jar:2.20.0-SNAPSHOT (compile) > com.alibaba:fastjson:jar:1.2.49 (compile) > org.apache.cxf:cxf-rt-transports-http:jar:3.1.2 (provided?) > org.apache.cxf:cxf-core:jar:3.1.2 (compile) > cglib:cglib-nodep:jar:2.2.2 (provided) > ant:ant:jar:1.6.2 (compile?) > xml-apis:xml-apis:jar:2.6.2 (compile?), maven-plugins:maven-cobertura-plugin:plugin:1.3 was not resolved. Dependency path: org.apache.beam:beam-sdks-java-extensions-sql-zetasql:jar:2.20.0-SNAPSHOT (compile) > org.apache.beam:beam-sdks-java-extensions-sql:jar:2.20.0-SNAPSHOT (compile) > com.alibaba:fastjson:jar:1.2.49 (compile) > org.apache.cxf:cxf-rt-transports-http:jar:3.1.2 (provided?) > org.apache.cxf:cxf-core:jar:3.1.2 (compile) > org.springframework:spring-beans:jar:4.1.6.RELEASE (compile?) > org.codehaus.groovy:groovy-all:jar:2.3.11 (compile?) > com.thoughtworks.xstream:xstream:jar:1.4.7 (compile?) > org.jdom:jdom:jar:1.1.3 (compile?) > jaxen:jaxen:jar:1.1.3 (compile?) > dom4j:dom4j:jar:1.6.1 (compile) > jaxen:jaxen:jar:1.1-beta-6 (compile?) > jdom:jdom:jar:1.0 (compile) > jaxen:jaxen:jar:1.0-FCS (compile?) > maven-plugins:maven-cobertura-plugin:plugin:1.3 (compile), maven-plugins:maven-findbugs-plugin:plugin:1.3.1 was not resolved. Dependency path: org.apache.beam:beam-sdks-java-extensions-sql-zetasql:jar:2.20.0-SNAPSHOT (compile) > org.apache.beam:beam-sdks-java-extensions-sql:jar:2.20.0-SNAPSHOT (compile) > com.alibaba:fastjson:jar:1.2.49 (compile) > org.apache.cxf:cxf-rt-transports-http:jar:3.1.2 (provided?) > org.apache.cxf:cxf-core:jar:3.1.2 (compile) > org.springframework:spring-beans:jar:4.1.6.RELEASE (compile?) > org.codehaus.groovy:groovy-all:jar:2.3.11 (compile?) > com.thoughtworks.xstream:xstream:jar:1.4.7 (compile?) > org.jdom:jdom:jar:1.1.3 (compile?) > jaxen:jaxen:jar:1.1.3 (compile?) > dom4j:dom4j:jar:1.6.1 (compile) > jaxen:jaxen:jar:1.1-beta-6 (compile?) > jdom:jdom:jar:1.0 (compile) > jaxen:jaxen:jar:1.0-FCS (compile?) > maven-plugins:maven-findbugs-plugin:plugin:1.3.1 (compile), com.oracle.oc4j:com.springsource.oracle.classloader:jar:10.1.3.1 was not resolved. Dependency path: org.apache.beam:beam-sdks-java-extensions-sql-zetasql:jar:2.20.0-SNAPSHOT (compile) > org.apache.beam:beam-sdks-java-extensions-sql:jar:2.20.0-SNAPSHOT (compile) > com.alibaba:fastjson:jar:1.2.49 (compile) > org.apache.cxf:cxf-rt-transports-http:jar:3.1.2 (provided?) > org.apache.cxf:cxf-core:jar:3.1.2 (compile) > org.springframework.osgi:spring-osgi-core:jar:1.2.1 (provided?) > org.springframework:org.springframework.context:jar:2.5.6.SEC01 (compile) > com.oracle.oc4j:com.springsource.oracle.classloader:jar:10.1.3.1 (provided), com.oracle.jdbc:com.springsource.oracle.jdbc:jar:10.2.0.2 was not resolved. Dependency path: org.apache.beam:beam-sdks-java-extensions-sql-zetasql:jar:2.20.0-SNAPSHOT (compile) > org.apache.beam:beam-sdks-java-extensions-sql:jar:2.20.0-SNAPSHOT (compile) > com.alibaba:fastjson:jar:1.2.49 (compile) > org.apache.cxf:cxf-rt-transports-http:jar:3.1.2 (provided?) > org.apache.cxf:cxf-core:jar:3.1.2 (compile) > org.springframework.osgi:spring-osgi-core:jar:1.2.1 (provided?) > org.springframework:org.springframework.test:jar:2.5.6.SEC01 (provided) > org.springframework:org.springframework.jdbc:jar:2.5.6.SEC01 (compile?) > com.experlog.xapool:com.springsource.org.enhydra.jdbc:jar:1.5.0 (compile?) > com.oracle.jdbc:com.springsource.oracle.jdbc:jar:10.2.0.2 (compile?), com.oracle.toplink:com.springsource.oracle.toplink:jar:10.1.3 was not resolved. Dependency path: org.apache.beam:beam-sdks-java-extensions-sql-zetasql:jar:2.20.0-SNAPSHOT (compile) > org.apache.beam:beam-sdks-java-extensions-sql:jar:2.20.0-SNAPSHOT (compile) > com.alibaba:fastjson:jar:1.2.49 (compile) > org.apache.cxf:cxf-rt-transports-http:jar:3.1.2 (provided?) > org.apache.cxf:cxf-core:jar:3.1.2 (compile) > org.springframework.osgi:spring-osgi-core:jar:1.2.1 (provided?) > org.springframework:org.springframework.test:jar:2.5.6.SEC01 (provided) > org.springframework:org.springframework.orm:jar:2.5.6.SEC01 (compile?) > com.oracle.toplink:com.springsource.oracle.toplink:jar:10.1.3 (compile?), com.oracle.aq:com.springsource.oracle.aq:jar:10.1.0.1 was not resolved. Dependency path: org.apache.beam:beam-sdks-java-extensions-sql-zetasql:jar:2.20.0-SNAPSHOT (compile) > org.apache.beam:beam-sdks-java-extensions-sql:jar:2.20.0-SNAPSHOT (compile) > com.alibaba:fastjson:jar:1.2.49 (compile) > org.apache.cxf:cxf-rt-transports-http:jar:3.1.2 (provided?) > org.apache.cxf:cxf-core:jar:3.1.2 (compile) > org.springframework.osgi:spring-osgi-core:jar:1.2.1 (provided?) > org.springframework:org.springframework.test:jar:2.5.6.SEC01 (provided) > org.springframework:org.springframework.orm:jar:2.5.6.SEC01 (compile?) > org.eclipse.persistence:com.springsource.org.eclipse.persistence:jar:1.0.0.m5 (compile?) > com.oracle.aq:com.springsource.oracle.aq:jar:10.1.0.1 (compile?), com.oracle.spatial:com.springsource.oracle.spatial.geometry:jar:10.2.0.1 was not resolved. Dependency path: org.apache.beam:beam-sdks-java-extensions-sql-zetasql:jar:2.20.0-SNAPSHOT (compile) > org.apache.beam:beam-sdks-java-extensions-sql:jar:2.20.0-SNAPSHOT (compile) > com.alibaba:fastjson:jar:1.2.49 (compile) > org.apache.cxf:cxf-rt-transports-http:jar:3.1.2 (provided?) > org.apache.cxf:cxf-core:jar:3.1.2 (compile) > org.springframework.osgi:spring-osgi-core:jar:1.2.1 (provided?) > org.springframework:org.springframework.test:jar:2.5.6.SEC01 (provided) > org.springframework:org.springframework.orm:jar:2.5.6.SEC01 (compile?) > org.eclipse.persistence:com.springsource.org.eclipse.persistence:jar:1.0.0.m5 (compile?) > com.oracle.spatial:com.springsource.oracle.spatial.geometry:jar:10.2.0.1 (compile?), com.oracle.xdb:com.springsource.oracle.xdb:jar:10.2.0.2 was not resolved. Dependency path: org.apache.beam:beam-sdks-java-extensions-sql-zetasql:jar:2.20.0-SNAPSHOT (compile) > org.apache.beam:beam-sdks-java-extensions-sql:jar:2.20.0-SNAPSHOT (compile) > com.alibaba:fastjson:jar:1.2.49 (compile) > org.apache.cxf:cxf-rt-transports-http:jar:3.1.2 (provided?) > org.apache.cxf:cxf-core:jar:3.1.2 (compile) > org.springframework.osgi:spring-osgi-core:jar:1.2.1 (provided?) > org.springframework:org.springframework.test:jar:2.5.6.SEC01 (provided) > org.springframework:org.springframework.orm:jar:2.5.6.SEC01 (compile?) > org.eclipse.persistence:com.springsource.org.eclipse.persistence:jar:1.0.0.m5 (compile?) > com.oracle.xdb:com.springsource.oracle.xdb:jar:10.2.0.2 (compile?), com.oracle.xml:com.springsource.oracle.xml:jar:10.2.0.2 was not resolved. Dependency path: org.apache.beam:beam-sdks-java-extensions-sql-zetasql:jar:2.20.0-SNAPSHOT (compile) > org.apache.beam:beam-sdks-java-extensions-sql:jar:2.20.0-SNAPSHOT (compile) > com.alibaba:fastjson:jar:1.2.49 (compile) > org.apache.cxf:cxf-rt-transports-http:jar:3.1.2 (provided?) > org.apache.cxf:cxf-core:jar:3.1.2 (compile) > org.springframework.osgi:spring-osgi-core:jar:1.2.1 (provided?) > org.springframework:org.springframework.test:jar:2.5.6.SEC01 (provided) > org.springframework:org.springframework.orm:jar:2.5.6.SEC01 (compile?) > org.eclipse.persistence:com.springsource.org.eclipse.persistence:jar:1.0.0.m5 (compile?) > com.oracle.xml:com.springsource.oracle.xml:jar:10.2.0.2 (compile?), com.jcraft:jsch.agentproxy:jar:0.0.6 was not resolved. Dependency path: org.apache.beam:beam-sdks-java-extensions-sql-zetasql:jar:2.20.0-SNAPSHOT (compile) > org.apache.beam:beam-sdks-java-extensions-sql:jar:2.20.0-SNAPSHOT (compile) > com.alibaba:fastjson:jar:1.2.49 (compile) > org.springframework:spring-websocket:jar:4.3.7.RELEASE (provided?) > org.springframework:spring-context:jar:4.3.7.RELEASE (compile) > org.codehaus.groovy:groovy-all:jar:2.4.9 (compile?) > org.apache.ivy:ivy:jar:2.4.0 (compile?) > com.jcraft:jsch.agentproxy:jar:0.0.6 (compile?), javax.security:jaas:jar:1.0.01 was not resolved. Dependency path: org.apache.beam:beam-sdks-java-extensions-sql-zetasql:jar:2.20.0-SNAPSHOT (compile) > org.apache.beam:beam-sdks-java-extensions-sql:jar:2.20.0-SNAPSHOT (compile) > com.alibaba:fastjson:jar:1.2.49 (compile) > org.springframework:spring-websocket:jar:4.3.7.RELEASE (provided?) > org.springframework:spring-webmvc:jar:4.3.7.RELEASE (compile?) > net.sf.jasperreports:jasperreports:jar:6.2.1 (compile?) > org.hibernate:hibernate-core:jar:3.3.2.GA (compile?) > javax.security:jaas:jar:1.0.01 (provided), javax.security:jacc:jar:1.0 was not resolved. Dependency path: org.apache.beam:beam-sdks-java-extensions-sql-zetasql:jar:2.20.0-SNAPSHOT (compile) > org.apache.beam:beam-sdks-java-extensions-sql:jar:2.20.0-SNAPSHOT (compile) > com.alibaba:fastjson:jar:1.2.49 (compile) > org.springframework:spring-websocket:jar:4.3.7.RELEASE (provided?) > org.springframework:spring-webmvc:jar:4.3.7.RELEASE (compile?) > net.sf.jasperreports:jasperreports:jar:6.2.1 (compile?) > org.hibernate:hibernate-core:jar:3.3.2.GA (compile?) > javax.security:jacc:jar:1.0 (provided), com.ibm.websphere:uow:jar:6.0.2.17 was not resolved. Dependency path: org.apache.beam:beam-sdks-java-extensions-sql-zetasql:jar:2.20.0-SNAPSHOT (compile) > org.apache.beam:beam-sdks-java-extensions-sql:jar:2.20.0-SNAPSHOT (compile) > com.alibaba:fastjson:jar:1.2.49 (compile) > org.springframework:spring-websocket:jar:4.3.7.RELEASE (provided?) > org.springframework:spring-webmvc:jar:4.3.7.RELEASE (compile?) > org.springframework:spring-context-support:jar:4.3.7.RELEASE (compile?) > org.springframework:spring-jdbc:jar:4.3.7.RELEASE (compile?) > org.springframework:spring-tx:jar:4.3.7.RELEASE (compile) > com.ibm.websphere:uow:jar:6.0.2.17 (compile?), org.jruby:readline:jar:1.0 was not resolved. Dependency path: org.apache.beam:beam-sdks-java-extensions-sql-zetasql:jar:2.20.0-SNAPSHOT (compile) > org.apache.beam:beam-sdks-java-extensions-sql:jar:2.20.0-SNAPSHOT (compile) > com.alibaba:fastjson:jar:1.2.49 (compile) > org.springframework.data:spring-data-redis:jar:1.8.6.RELEASE (provided?) > org.springframework.data:spring-data-keyvalue:jar:1.2.6.RELEASE (compile) > org.springframework.data:spring-data-commons:jar:1.13.6.RELEASE (compile) > org.springframework.hateoas:spring-hateoas:jar:0.23.0.RELEASE (compile?) > org.springframework:spring-aop:jar:4.3.5.RELEASE (compile) > com.jamonapi:jamon:jar:2.81 (compile?) > org.springframework:spring-context:jar:4.0.2.RELEASE (provided) > org.jruby:jruby:jar:1.7.10 (compile?) > org.jruby:jruby-stdlib:jar:1.7.10 (compile) > org.jruby:readline:jar:1.0 (provided), org.jruby:openssl:jar:0.9.4 was not resolved. Dependency path: org.apache.beam:beam-sdks-java-extensions-sql-zetasql:jar:2.20.0-SNAPSHOT (compile) > org.apache.beam:beam-sdks-java-extensions-sql:jar:2.20.0-SNAPSHOT (compile) > com.alibaba:fastjson:jar:1.2.49 (compile) > org.springframework.data:spring-data-redis:jar:1.8.6.RELEASE (provided?) > org.springframework.data:spring-data-keyvalue:jar:1.2.6.RELEASE (compile) > org.springframework.data:spring-data-commons:jar:1.13.6.RELEASE (compile) > org.springframework.hateoas:spring-hateoas:jar:0.23.0.RELEASE (compile?) > org.springframework:spring-aop:jar:4.3.5.RELEASE (compile) > com.jamonapi:jamon:jar:2.81 (compile?) > org.springframework:spring-context:jar:4.0.2.RELEASE (provided) > org.jruby:jruby:jar:1.7.10 (compile?) > org.jruby:jruby-stdlib:jar:1.7.10 (compile) > org.jruby:openssl:jar:0.9.4 (provided)
        at com.google.cloud.tools.opensource.classpath.LinkageCheckerArguments.getInputClasspath(LinkageCheckerArguments.java:209)
        at com.google.cloud.tools.opensource.classpath.LinkageCheckerArguments.getEntryPointJars(LinkageCheckerArguments.java:230)
        at com.google.cloud.tools.opensource.classpath.LinkageCheckerMain.main(LinkageCheckerMain.java:74)

> Task :checkJavaLinkage FAILED

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':checkJavaLinkage'.
> Process 'command '/usr/lib/jvm/java-8-openjdk-amd64/bin/java'' finished with non-zero exit value 1

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.

* Get more help at https://help.gradle.org

Deprecated Gradle features were used in this build, making it incompatible with Gradle 6.0.
Use '--warning-mode all' to show the individual deprecation warnings.
See https://docs.gradle.org/5.2.1/userguide/command_line_interface.html#sec:command_line_warnings

BUILD FAILED in 6m 54s
906 actionable tasks: 622 executed, 284 up-to-date

```

--- a/dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/LinkageCheckerArguments.java
+++ b/dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/LinkageCheckerArguments.java
@@ -20,8 +20,6 @@ import static com.google.common.base.Preconditions.checkNotNull;
 import static com.google.common.collect.ImmutableList.toImmutableList;
 
 import com.google.cloud.tools.opensource.dependencies.RepositoryUtility;
-import com.google.cloud.tools.opensource.dependencies.UnresolvableArtifactProblem;
-import com.google.common.base.Joiner;
 import com.google.common.collect.ImmutableList;
 import com.google.common.collect.ImmutableSet;
 import java.nio.file.Path;
@@ -60,7 +58,7 @@ final class LinkageCheckerArguments {
   private static final HelpFormatter helpFormatter = new HelpFormatter();
 
   private final CommandLine commandLine;
-  private ImmutableList<Path> cachedInputClasspath;
+  private ClassPathResult classPathResult;
   private ImmutableList<Artifact> cachedArtifacts;
   private final ImmutableList<String> extraMavenRepositoryUrls;
   private final boolean addMavenCentral;
@@ -191,33 +189,34 @@ final class LinkageCheckerArguments {
   }
 
   /**
+   * Returns class path result for a BOM or Maven artifacts resolution. Null if the input is a list
+   * of JAR files.
+   */
+  ClassPathResult getClassPathResult() {
+    return classPathResult;
+  }
+
+  /**
    * Returns a list of absolute paths to jar files for the option. The list includes dependencies if
    * BOM or Maven coordinates are specified in the option.
    */
   ImmutableList<Path> getInputClasspath() throws RepositoryException {
-    if (cachedInputClasspath != null) {
+    if (classPathResult != null) {
       // Avoid unnecessary dependency resolution, which Cloud OSS BOM takes around 4 seconds
-      return cachedInputClasspath;
+      return classPathResult.getClassPath();
     }
 
     if (commandLine.hasOption("b") || commandLine.hasOption("a")) {
       List<Artifact> artifacts = getArtifacts();
-      ClassPathResult result = classPathBuilder.resolve(artifacts);
-      ImmutableList<UnresolvableArtifactProblem> artifactProblems = result.getArtifactProblems();
-      if (!artifactProblems.isEmpty()) {
-        throw new RepositoryException(
-            "Unresolved artifacts: " + Joiner.on(", ").join(artifactProblems));
-      }
-      cachedInputClasspath = result.getClassPath();
+      classPathResult = classPathBuilder.resolve(artifacts);
+      return classPathResult.getClassPath();
     } else {
       // b, a, or j is specified in OptionGroup
       String[] jarFiles = commandLine.getOptionValues("j");
-      cachedInputClasspath =
-          Arrays.stream(jarFiles)
-              .map(name -> Paths.get(name).toAbsolutePath())
-              .collect(toImmutableList());
+      return Arrays.stream(jarFiles)
+          .map(name -> Paths.get(name).toAbsolutePath())
+          .collect(toImmutableList());
     }
-    return cachedInputClasspath;
   }
 
   /** Returns a set of jar files that hold entry point classes. */

--- a/dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/LinkageCheckerArguments.java
+++ b/dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/LinkageCheckerArguments.java
@@ -21,11 +21,9 @@ import static com.google.common.collect.ImmutableList.toImmutableList;
 
 import com.google.cloud.tools.opensource.dependencies.RepositoryUtility;
 import com.google.common.collect.ImmutableList;
-import com.google.common.collect.ImmutableSet;
 import java.nio.file.Path;
 import java.nio.file.Paths;
 import java.util.Arrays;
-import java.util.List;
 import org.apache.commons.cli.CommandLine;
 import org.apache.commons.cli.CommandLineParser;
 import org.apache.commons.cli.DefaultParser;
@@ -58,12 +56,10 @@ final class LinkageCheckerArguments {
   private static final HelpFormatter helpFormatter = new HelpFormatter();
 
   private final CommandLine commandLine;
-  private ClassPathResult classPathResult;
   private ImmutableList<Artifact> cachedArtifacts;
   private final ImmutableList<String> extraMavenRepositoryUrls;
   private final boolean addMavenCentral;
   private final boolean reportOnlyReachable;
-  private final ClassPathBuilder classPathBuilder = new ClassPathBuilder();
 
   private LinkageCheckerArguments(CommandLine commandLine) {
     this.commandLine = checkNotNull(commandLine);
@@ -188,48 +184,15 @@ final class LinkageCheckerArguments {
     }
   }
 
-  /**
-   * Returns class path result for a BOM or Maven artifacts resolution. Null if the input is a list
-   * of JAR files.
-   */
-  ClassPathResult getClassPathResult() {
-    return classPathResult;
-  }
-
-  /**
-   * Returns a list of absolute paths to jar files for the option. The list includes dependencies if
-   * BOM or Maven coordinates are specified in the option.
-   */
-  ImmutableList<Path> getInputClasspath() throws RepositoryException {
-    if (classPathResult != null) {
-      // Avoid unnecessary dependency resolution, which Cloud OSS BOM takes around 4 seconds
-      return classPathResult.getClassPath();
-    }
-
-    if (commandLine.hasOption("b") || commandLine.hasOption("a")) {
-      List<Artifact> artifacts = getArtifacts();
-      classPathResult = classPathBuilder.resolve(artifacts);
-      return classPathResult.getClassPath();
-    } else {
-      // b, a, or j is specified in OptionGroup
+  /** Returns a list of absolute paths to files specified in the JAR file option. */
+  ImmutableList<Path> getJarFiles() {
+    if (commandLine.hasOption("j")) {
       String[] jarFiles = commandLine.getOptionValues("j");
       return Arrays.stream(jarFiles)
           .map(name -> Paths.get(name).toAbsolutePath())
           .collect(toImmutableList());
-    }
-  }
-
-  /** Returns a set of jar files that hold entry point classes. */
-  ImmutableSet<Path> getEntryPointJars() throws RepositoryException {
-    if (commandLine.hasOption("a") || commandLine.hasOption('b')) {
-      // For an artifact list (or a BOM), the first elements in inputClasspath are the artifacts
-      // specified the list, followed by their dependencies.
-      int artifactCount = getArtifacts().size();
-      // For Maven artifact list (or a BOM), entry point classes are ones in the list
-      return ImmutableSet.copyOf(getInputClasspath().subList(0, artifactCount));
     } else {
-      // For list of jar files, entry point classes are all classes in the files
-      return ImmutableSet.copyOf(getInputClasspath());
+      throw new IllegalArgumentException("The arguments must have option 'j' to list JAR files");
     }
   }
 

--- a/dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/LinkageCheckerMain.java
+++ b/dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/LinkageCheckerMain.java
@@ -53,25 +53,30 @@ class LinkageCheckerMain {
     DependencyGraphBuilder dependencyGraphBuilder =
         new DependencyGraphBuilder(linkageCheckerArguments.getMavenRepositoryUrls());
 
-    ClassPathBuilder classPathBuilder = new ClassPathBuilder(dependencyGraphBuilder);
+    // This is non-empty if a BOM or artifacts are specified in the argument
     ImmutableList<Artifact> artifacts = linkageCheckerArguments.getArtifacts();
 
     // When JAR files are specified in the argument, artifacts are empty.
     ImmutableList<Path> inputClassPath;
     List<ArtifactProblem> artifactProblems = new ArrayList<>();
     ClassPathResult classPathResult = null;
+
+    ImmutableSet<Path> entryPointJars;
     if (artifacts.isEmpty()) {
       // When JAR files are passed as arguments, classPathResult is null, because there is no need
       // to resolve Maven dependencies.
-      inputClassPath = linkageCheckerArguments.getInputClasspath();
+      inputClassPath = linkageCheckerArguments.getJarFiles();
+      entryPointJars = ImmutableSet.copyOf(inputClassPath);
     } else {
+      // When a BOM or Maven artifacts are passed as arguments, resolve the dependencies.
+      ClassPathBuilder classPathBuilder = new ClassPathBuilder(dependencyGraphBuilder);
+      classPathResult = classPathBuilder.resolve(artifacts);
       // When Maven artifacts (or a BOM) are passed as arguments, resolve the dependency tree.
-      inputClassPath = linkageCheckerArguments.getInputClasspath();
-      classPathResult = linkageCheckerArguments.getClassPathResult();
+      inputClassPath = classPathResult.getClassPath();
       artifactProblems.addAll(classPathResult.getArtifactProblems());
+      entryPointJars = ImmutableSet.copyOf(inputClassPath.subList(0, artifacts.size()));
     }
 
-    ImmutableSet<Path> entryPointJars = linkageCheckerArguments.getEntryPointJars();
     LinkageChecker linkageChecker = LinkageChecker.create(inputClassPath, entryPointJars);
     ImmutableSetMultimap<SymbolProblem, ClassFile> symbolProblems =
         linkageChecker.findSymbolProblems();

--- a/dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/LinkageCheckerMain.java
+++ b/dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/LinkageCheckerMain.java
@@ -66,8 +66,8 @@ class LinkageCheckerMain {
       inputClassPath = linkageCheckerArguments.getInputClasspath();
     } else {
       // When Maven artifacts (or a BOM) are passed as arguments, resolve the dependency tree.
-      classPathResult = classPathBuilder.resolve(artifacts);
-      inputClassPath = classPathResult.getClassPath();
+      inputClassPath = linkageCheckerArguments.getInputClasspath();
+      classPathResult = linkageCheckerArguments.getClassPathResult();
       artifactProblems.addAll(classPathResult.getArtifactProblems());
     }
 

--- a/dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/LinkageCheckerMain.java
+++ b/dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/LinkageCheckerMain.java
@@ -50,18 +50,16 @@ class LinkageCheckerMain {
     LinkageCheckerArguments linkageCheckerArguments =
         LinkageCheckerArguments.readCommandLine(arguments);
 
-    DependencyGraphBuilder dependencyGraphBuilder =
-        new DependencyGraphBuilder(linkageCheckerArguments.getMavenRepositoryUrls());
-
     // This is non-empty if a BOM or artifacts are specified in the argument
     ImmutableList<Artifact> artifacts = linkageCheckerArguments.getArtifacts();
 
     // When JAR files are specified in the argument, artifacts are empty.
     ImmutableList<Path> inputClassPath;
+    ImmutableSet<Path> entryPointJars;
     List<ArtifactProblem> artifactProblems = new ArrayList<>();
+    // classPathResult is kept null if JAR files are specified in the argument
     ClassPathResult classPathResult = null;
 
-    ImmutableSet<Path> entryPointJars;
     if (artifacts.isEmpty()) {
       // When JAR files are passed as arguments, classPathResult is null, because there is no need
       // to resolve Maven dependencies.
@@ -69,9 +67,10 @@ class LinkageCheckerMain {
       entryPointJars = ImmutableSet.copyOf(inputClassPath);
     } else {
       // When a BOM or Maven artifacts are passed as arguments, resolve the dependencies.
+      DependencyGraphBuilder dependencyGraphBuilder =
+          new DependencyGraphBuilder(linkageCheckerArguments.getMavenRepositoryUrls());
       ClassPathBuilder classPathBuilder = new ClassPathBuilder(dependencyGraphBuilder);
       classPathResult = classPathBuilder.resolve(artifacts);
-      // When Maven artifacts (or a BOM) are passed as arguments, resolve the dependency tree.
       inputClassPath = classPathResult.getClassPath();
       artifactProblems.addAll(classPathResult.getArtifactProblems());
       entryPointJars = ImmutableSet.copyOf(inputClassPath.subList(0, artifacts.size()));

--- a/dependencies/src/test/java/com/google/cloud/tools/opensource/classpath/LinkageCheckerArgumentsTest.java
+++ b/dependencies/src/test/java/com/google/cloud/tools/opensource/classpath/LinkageCheckerArgumentsTest.java
@@ -120,6 +120,37 @@ public class LinkageCheckerArgumentsTest {
   }
 
   @Test
+  public void testGetClassPathResult_jarFiles() throws ParseException, RepositoryException {
+    LinkageCheckerArguments parsedArguments =
+        LinkageCheckerArguments.readCommandLine("-j", "dummy.jar", "-r");
+
+    parsedArguments.getInputClasspath();
+
+    // There's no Maven dependency resolution for list of JAR files
+    Assert.assertNull(parsedArguments.getClassPathResult());
+  }
+
+  @Test
+  public void testGetClassPathResult_mavenArtifacts() throws ParseException, RepositoryException {
+    LinkageCheckerArguments parsedArgumentsForArtifact =
+        LinkageCheckerArguments.readCommandLine("-a", "ant:ant:1.6.2");
+
+    parsedArgumentsForArtifact.getInputClasspath();
+
+    Assert.assertNotNull(parsedArgumentsForArtifact.getClassPathResult());
+  }
+
+  @Test
+  public void testGetClassPathResult_bom() throws ParseException, RepositoryException {
+    LinkageCheckerArguments parsedArgumentsForArtifact =
+        LinkageCheckerArguments.readCommandLine("-b", "com.google.cloud:libraries-bom:1.0.0");
+
+    parsedArgumentsForArtifact.getInputClasspath();
+
+    Assert.assertNotNull(parsedArgumentsForArtifact.getClassPathResult());
+  }
+
+  @Test
   public void testGetInputClasspath_shouldNotThrowException()
       throws ParseException, RepositoryException {
     LinkageCheckerArguments parsedArguments =

--- a/dependencies/src/test/java/com/google/cloud/tools/opensource/classpath/LinkageCheckerArgumentsTest.java
+++ b/dependencies/src/test/java/com/google/cloud/tools/opensource/classpath/LinkageCheckerArgumentsTest.java
@@ -16,9 +16,11 @@
 
 package com.google.cloud.tools.opensource.classpath;
 
-import com.google.common.collect.ImmutableList;
+import static com.google.cloud.tools.opensource.classpath.ClassPathBuilderTest.PATH_FILE_NAMES;
+
 import com.google.common.truth.Truth;
 import java.nio.file.Path;
+import java.util.List;
 import org.apache.commons.cli.ParseException;
 import org.eclipse.aether.RepositoryException;
 import org.junit.Assert;
@@ -59,11 +61,39 @@ public class LinkageCheckerArgumentsTest {
   }
 
   @Test
-  public void testReadCommandLine_multipleJars() throws ParseException, RepositoryException {
+  public void testReadCommandLine_multipleJars() throws ParseException {
     LinkageCheckerArguments parsedArguments =
         LinkageCheckerArguments.readCommandLine(
             "-j", "/foo/bar/A.jar,/foo/bar/B.jar,/foo/bar/C.jar");
-    Truth.assertThat(parsedArguments.getInputClasspath()).hasSize(3);
+
+    Truth.assertThat(parsedArguments.getJarFiles()).hasSize(3);
+  }
+
+  @Test
+  public void testgetJarFiles_jarFileList() throws ParseException {
+
+    LinkageCheckerArguments parsedArguments =
+        LinkageCheckerArguments.readCommandLine("--jars", "dir1/foo.jar,dir2/bar.jar,baz.jar");
+    List<Path> inputClasspath = parsedArguments.getJarFiles();
+
+    Truth.assertThat(inputClasspath)
+        .comparingElementsUsing(PATH_FILE_NAMES)
+        .containsExactly("foo.jar", "bar.jar", "baz.jar");
+  }
+
+  @Test
+  public void testgetJarFiles_invalidOption() throws ParseException {
+    LinkageCheckerArguments parsedArguments =
+        LinkageCheckerArguments.readCommandLine(
+            "--artifacts", "com.google.guava:guava:26.0,io.grpc:grpc-core:1.17.1");
+
+    try {
+      parsedArguments.getJarFiles();
+      Assert.fail();
+    } catch (IllegalArgumentException ex) {
+      // pass
+      Assert.assertEquals("The arguments must have option 'j' to list JAR files", ex.getMessage());
+    }
   }
 
   @Test
@@ -117,47 +147,5 @@ public class LinkageCheckerArgumentsTest {
         LinkageCheckerArguments.readCommandLine("-j", "dummy.jar", "-r");
 
     Truth.assertThat(parsedArguments.getReportOnlyReachable()).isTrue();
-  }
-
-  @Test
-  public void testGetClassPathResult_jarFiles() throws ParseException, RepositoryException {
-    LinkageCheckerArguments parsedArguments =
-        LinkageCheckerArguments.readCommandLine("-j", "dummy.jar", "-r");
-
-    parsedArguments.getInputClasspath();
-
-    // There's no Maven dependency resolution for list of JAR files
-    Assert.assertNull(parsedArguments.getClassPathResult());
-  }
-
-  @Test
-  public void testGetClassPathResult_mavenArtifacts() throws ParseException, RepositoryException {
-    LinkageCheckerArguments parsedArgumentsForArtifact =
-        LinkageCheckerArguments.readCommandLine("-a", "ant:ant:1.6.2");
-
-    parsedArgumentsForArtifact.getInputClasspath();
-
-    Assert.assertNotNull(parsedArgumentsForArtifact.getClassPathResult());
-  }
-
-  @Test
-  public void testGetClassPathResult_bom() throws ParseException, RepositoryException {
-    LinkageCheckerArguments parsedArgumentsForArtifact =
-        LinkageCheckerArguments.readCommandLine("-b", "com.google.cloud:libraries-bom:1.0.0");
-
-    parsedArgumentsForArtifact.getInputClasspath();
-
-    Assert.assertNotNull(parsedArgumentsForArtifact.getClassPathResult());
-  }
-
-  @Test
-  public void testGetInputClasspath_shouldNotThrowException()
-      throws ParseException, RepositoryException {
-    LinkageCheckerArguments parsedArguments =
-        LinkageCheckerArguments.readCommandLine("-a", "ant:ant:1.6.2");
-
-    // This should not raise an exception
-    ImmutableList<Path> inputClasspath = parsedArguments.getInputClasspath();
-    Assert.assertEquals("ant-1.6.2.jar", inputClasspath.get(0).getFileName().toString());
   }
 }

--- a/dependencies/src/test/java/com/google/cloud/tools/opensource/classpath/LinkageCheckerArgumentsTest.java
+++ b/dependencies/src/test/java/com/google/cloud/tools/opensource/classpath/LinkageCheckerArgumentsTest.java
@@ -16,7 +16,9 @@
 
 package com.google.cloud.tools.opensource.classpath;
 
+import com.google.common.collect.ImmutableList;
 import com.google.common.truth.Truth;
+import java.nio.file.Path;
 import org.apache.commons.cli.ParseException;
 import org.eclipse.aether.RepositoryException;
 import org.junit.Assert;
@@ -115,5 +117,16 @@ public class LinkageCheckerArgumentsTest {
         LinkageCheckerArguments.readCommandLine("-j", "dummy.jar", "-r");
 
     Truth.assertThat(parsedArguments.getReportOnlyReachable()).isTrue();
+  }
+
+  @Test
+  public void testGetInputClasspath_shouldNotThrowException()
+      throws ParseException, RepositoryException {
+    LinkageCheckerArguments parsedArguments =
+        LinkageCheckerArguments.readCommandLine("-a", "ant:ant:1.6.2");
+
+    // This should not raise an exception
+    ImmutableList<Path> inputClasspath = parsedArguments.getInputClasspath();
+    Assert.assertEquals("ant-1.6.2.jar", inputClasspath.get(0).getFileName().toString());
   }
 }

--- a/dependencies/src/test/java/com/google/cloud/tools/opensource/classpath/LinkageCheckerArgumentsTest.java
+++ b/dependencies/src/test/java/com/google/cloud/tools/opensource/classpath/LinkageCheckerArgumentsTest.java
@@ -66,11 +66,13 @@ public class LinkageCheckerArgumentsTest {
         LinkageCheckerArguments.readCommandLine(
             "-j", "/foo/bar/A.jar,/foo/bar/B.jar,/foo/bar/C.jar");
 
-    Truth.assertThat(parsedArguments.getJarFiles()).hasSize(3);
+    Truth.assertThat(parsedArguments.getJarFiles())
+        .comparingElementsUsing(PATH_FILE_NAMES)
+        .containsExactly("A.jar", "B.jar", "C.jar");
   }
 
   @Test
-  public void testgetJarFiles_jarFileList() throws ParseException {
+  public void testGetJarFiles_jarFileList() throws ParseException {
 
     LinkageCheckerArguments parsedArguments =
         LinkageCheckerArguments.readCommandLine("--jars", "dir1/foo.jar,dir2/bar.jar,baz.jar");
@@ -82,7 +84,7 @@ public class LinkageCheckerArgumentsTest {
   }
 
   @Test
-  public void testgetJarFiles_invalidOption() throws ParseException {
+  public void testGetJarFiles_invalidOption() throws ParseException {
     LinkageCheckerArguments parsedArguments =
         LinkageCheckerArguments.readCommandLine(
             "--artifacts", "com.google.guava:guava:26.0,io.grpc:grpc-core:1.17.1");

--- a/dependencies/src/test/java/com/google/cloud/tools/opensource/classpath/LinkageCheckerTest.java
+++ b/dependencies/src/test/java/com/google/cloud/tools/opensource/classpath/LinkageCheckerTest.java
@@ -25,8 +25,10 @@ import static org.junit.Assert.assertNotNull;
 import static org.junit.Assert.assertSame;
 import static org.junit.Assert.assertTrue;
 
+import com.google.cloud.tools.opensource.dependencies.ArtifactProblem;
 import com.google.cloud.tools.opensource.dependencies.DependencyGraph;
 import com.google.cloud.tools.opensource.dependencies.DependencyGraphBuilder;
+import com.google.cloud.tools.opensource.dependencies.UnresolvableArtifactProblem;
 import com.google.common.collect.ImmutableList;
 import com.google.common.collect.ImmutableSet;
 import com.google.common.collect.ImmutableSetMultimap;
@@ -666,8 +668,8 @@ public class LinkageCheckerTest {
   }
 
   @Test
-  public void testGenerateInputClasspathFromLinkageCheckOption_failOnMissingDependency()
-      throws ParseException {
+  public void testGenerateInputClasspathFromLinkageCheckOption_recordMissingDependency()
+      throws ParseException, RepositoryException {
     // tomcat-jasper has missing dependency (not optional):
     //   org.apache.tomcat:tomcat-jasper:jar:8.0.9
     //     org.eclipse.jdt.core.compiler:ecj:jar:4.4RC4 (not found in Maven central)
@@ -675,14 +677,18 @@ public class LinkageCheckerTest {
         LinkageCheckerArguments.readCommandLine(
             "--artifacts", "org.apache.tomcat:tomcat-jasper:8.0.9");
 
-    try {
-      parsedArguments.getInputClasspath();
-      Assert.fail(
-          "Because the unavailable dependency is not optional, it should throw an exception");
-    } catch (RepositoryException ex) {
-      Truth.assertThat(ex.getMessage())
-          .startsWith("Unresolved artifacts: org.eclipse.jdt.core.compiler:ecj:jar:4.4RC4");
-    }
+    parsedArguments.getInputClasspath();
+
+    ClassPathResult classPathResult = parsedArguments.getClassPathResult();
+    assertNotNull(classPathResult);
+    ImmutableList<UnresolvableArtifactProblem> artifactProblems =
+        classPathResult.getArtifactProblems();
+    Truth.assertThat(artifactProblems)
+        .comparingElementsUsing(
+            Correspondence.transforming(
+                (ArtifactProblem problem) -> problem.getArtifact().toString(),
+                "problem with Maven coordinate"))
+        .contains("org.eclipse.jdt.core.compiler:ecj:jar:4.4RC4");
   }
 
   @Test

--- a/dependencies/src/test/java/com/google/cloud/tools/opensource/classpath/LinkageCheckerTest.java
+++ b/dependencies/src/test/java/com/google/cloud/tools/opensource/classpath/LinkageCheckerTest.java
@@ -63,6 +63,7 @@ public class LinkageCheckerTest {
 
   private Path guavaPath;
   private Path firestorePath;
+  private ClassPathBuilder classPathBuilder = new ClassPathBuilder();
 
   /** Returns JAR files resolved for the full dependency tree of {@code coordinates}. */
   static ImmutableList<Path> resolvePaths(String... coordinates) {
@@ -600,7 +601,9 @@ public class LinkageCheckerTest {
 
     LinkageCheckerArguments parsedArguments =
         LinkageCheckerArguments.readCommandLine("-b", bomCoordinates);
-    ImmutableList<Path> inputClasspath = parsedArguments.getInputClasspath();
+    ImmutableList<Path> inputClasspath =
+        classPathBuilder.resolve(parsedArguments.getArtifacts()).getClassPath();
+
     Truth.assertThat(inputClasspath).isNotEmpty();
 
     List<String> names =
@@ -630,7 +633,8 @@ public class LinkageCheckerTest {
 
     LinkageCheckerArguments parsedArguments =
         LinkageCheckerArguments.readCommandLine("--artifacts", mavenCoordinates);
-    List<Path> inputClasspath = parsedArguments.getInputClasspath();
+    List<Path> inputClasspath =
+        classPathBuilder.resolve(parsedArguments.getArtifacts()).getClassPath();
 
     Truth.assertWithMessage(
             "The first 2 items in the classpath should be the 2 artifacts in the input")
@@ -659,7 +663,8 @@ public class LinkageCheckerTest {
     LinkageCheckerArguments parsedArguments =
         LinkageCheckerArguments.readCommandLine("--artifacts", "com.google.guava:guava-gwt:20.0");
 
-    ImmutableList<Path> inputClasspath = parsedArguments.getInputClasspath();
+    ImmutableList<Path> inputClasspath =
+        classPathBuilder.resolve(parsedArguments.getArtifacts()).getClassPath();
 
     Truth.assertThat(inputClasspath)
         .comparingElementsUsing(PATH_FILE_NAMES)
@@ -676,31 +681,14 @@ public class LinkageCheckerTest {
         LinkageCheckerArguments.readCommandLine(
             "--artifacts", "org.apache.tomcat:tomcat-jasper:8.0.9");
 
-    parsedArguments.getInputClasspath();
-
-    ClassPathResult classPathResult = parsedArguments.getClassPathResult();
-    assertNotNull(classPathResult);
     ImmutableList<UnresolvableArtifactProblem> artifactProblems =
-        classPathResult.getArtifactProblems();
+        classPathBuilder.resolve(parsedArguments.getArtifacts()).getArtifactProblems();
     Truth.assertThat(artifactProblems)
         .comparingElementsUsing(
             Correspondence.transforming(
                 (UnresolvableArtifactProblem problem) -> problem.getArtifact().toString(),
                 "problem with Maven coordinate"))
         .contains("org.eclipse.jdt.core.compiler:ecj:jar:4.4RC4");
-  }
-
-  @Test
-  public void testGenerateInputClasspath_jarFileList()
-      throws RepositoryException, ParseException {
-
-    LinkageCheckerArguments parsedArguments =
-        LinkageCheckerArguments.readCommandLine("--jars", "dir1/foo.jar,dir2/bar.jar,baz.jar");
-    List<Path> inputClasspath = parsedArguments.getInputClasspath();
-
-    Truth.assertThat(inputClasspath)
-        .comparingElementsUsing(PATH_FILE_NAMES)
-        .containsExactly("foo.jar", "bar.jar", "baz.jar");
   }
 
   @Test

--- a/dependencies/src/test/java/com/google/cloud/tools/opensource/classpath/LinkageCheckerTest.java
+++ b/dependencies/src/test/java/com/google/cloud/tools/opensource/classpath/LinkageCheckerTest.java
@@ -25,7 +25,6 @@ import static org.junit.Assert.assertNotNull;
 import static org.junit.Assert.assertSame;
 import static org.junit.Assert.assertTrue;
 
-import com.google.cloud.tools.opensource.dependencies.ArtifactProblem;
 import com.google.cloud.tools.opensource.dependencies.DependencyGraph;
 import com.google.cloud.tools.opensource.dependencies.DependencyGraphBuilder;
 import com.google.cloud.tools.opensource.dependencies.UnresolvableArtifactProblem;
@@ -686,7 +685,7 @@ public class LinkageCheckerTest {
     Truth.assertThat(artifactProblems)
         .comparingElementsUsing(
             Correspondence.transforming(
-                (ArtifactProblem problem) -> problem.getArtifact().toString(),
+                (UnresolvableArtifactProblem problem) -> problem.getArtifact().toString(),
                 "problem with Maven coordinate"))
         .contains("org.eclipse.jdt.core.compiler:ecj:jar:4.4RC4");
   }

